### DISCUSSION TITLE
How to Emit An Event Upon JupyterLab Fully Functional State

### DISCUSSION THREAD

**Discussion by thick-as-thieves** [jupyterlab/jupyterlab#16707] [Q&A]
Hello,

I am currently working on emitting an event to signify when JupyterLab has reached a fully functional state after page load. For this purpose, I am defining "fully functional" as the moment when the last opened resource (such as a notebook, markdown file, text file, etc.) is completely loaded.

Q1:
Is there an existing event or signal that is triggered when any resource-type file is loaded during JupyterLab's page load? If not, how can I emit such a signal?

Q2:
Additionally, is there a better approach to determine when JupyterLab has reached a fully functional state?

I would appreciate your guidance on these questions.
Thank you!

**Comment by krassowski** [comment]
`app.shell.restored` is a signal emitted when the lab shell gets restored/fully loaded. Some extensions will use it to only then perform certain actions. It does not mean that e.g. the notebooks that are open get completed loaded, but just that their widgets would have been added to the main area.

It gets more complex when you look at the details because definition of "loaded" varies since JupyterLab uses lazy loading or partial rendering for many elements for performance reasons.
